### PR TITLE
documentation and minor changes

### PR DESCRIPTION
**Pull request by hen-ter**
Attempt to document the code.
I used indenting in the comments to make it easier to distinguish the different code segments.

Note that I swapped XCHG<=>AND in the food generation to have AND AX, CX for a fairer comparison with AND AX, 0xF9C.
I also changed JA to JAE in the vertical boundary check since I think that's clearer (DI=0xFA0 => game over)

Feel free to adjust everything to your liking.

One additional thought, how about making an alternative version with AND AX, 0xF9C so the game cannot be ruined by food spawning off screen and HLT to slow it down a bit? It's still only 78 bytes and personally I find it surprisingly fun that way :)

### CODE REVIEW

File: snake.asm [hunk 1]
```diff
@@ -1,41 +1,46 @@
...
+    div cl              ;   divide AX by 160 (width of a screen line)
+    and ah, ah          ;   check if remainder in AH is 0, which is the case only when head is on the left edge and movement was right (-4) or head is on the right edge and movement was left (4)
+    jz start            ;     if so => game over
+    cmp [di], ch        ; check if head position already contains snake
+    je start            ;   if so => game over
+    push di             ; push new head position onto the stack
+    cmp [di], cl        ; check if head position contains food item
+    mov [di], ch        ;   and place snake character onto screen
+    je .food            ; if food was consumed => generate new food
+    es lodsw            ; no food was consumed so load old tail position into AX and update SI to point to new tail position
```
Comment (donno2048) [inline snake.asm:43]: Maybe explain this only works because `si` points to the stack?

File: snake.asm [hunk 2]
```diff
@@ -1,41 +1,46 @@
...
-	cmp [di], cl
-	mov [di], ch
-	je .food
-	es lodsw
-	xchg ax, bx
-	mov [bx], ah
-	jmp SHORT .input
+; register usage during main loop
+; DS: 0xB800, segment of screen buffer
+; CX: 0xFA0, size of the screen (80x25x16bit). CH (0xF) is also used as the snake character, CL (0xA0) as the food character
```
Comment (donno2048) [inline snake.asm:3]: CL is also used as the width of the screen

**Review by donno2048 — COMMENTED** [review]
Awesome work! Never even thought the comments could be so explanatory!

**Comment by donno2048** [comment]
> One additional thought, how about making an alternative version with AND AX, 0xF9C so the game cannot be ruined by food spawning off screen and HLT to slow it down a bit? It's still only 78 bytes and personally I find it surprisingly fun that way :)

Actually thought about it but it feels weird to have a project aimed towards making something as small as I can but have a slightly bigger version which is just better at some other things... And actually not doing that is the only reason I'm not making a "super small version" and just allowing the snake to cross itself or food to spawn on it... But I don't know I might do it either way...

**Comment by donno2048** [comment]
Sorry for the delay by the way

**Comment by hen-ter** [comment]
> > One additional thought, how about making an alternative version with AND AX, 0xF9C so the game cannot be ruined by food spawning off screen and HLT to slow it down a bit? It's still only 78 bytes and personally I find it surprisingly fun that way :)
> 
> Actually thought about it but it feels weird to have a project aimed towards making something as small as I can but have a slightly bigger version which is just better at some other things... And actually not doing that is the only reason I'm not making a "super small version" and just allowing the snake to cross itself or food to spawn on it... But I don't know I might do it either way...

Totally get your point. The food off screen is kind of a similar issue though...

**Comment by hen-ter** [comment]
> Awesome work! Never even thought the comments could be so explanatory!

Thanks! It's much easier to be explanatory when not trying to strictly comment each line individually.

**Pull request merged**